controllers/volumereplicationgroup_controller.go

```diff
@@ -462,27 +468,132 @@ func (v *VRGInstance) createVolumeReplicationResources() bool {
 
 		volRep := &volrep.VolumeReplication{}
 
-		err := v.reconciler.Get(v.ctx, types.NamespacedName{Name: pvc.Name, Namespace: pvc.Namespace}, volRep)
-		if err != nil {
-			if errors.IsNotFound(err) {
-				log.Info("Creating VolumeReplication resource for PersistentVolumeClaim")
+		err := v.reconciler.Get(v.ctx, pvcNamespacedName, volRep)
+		if err != nil && errors.IsNotFound(err) {
+			log.Info("Uploading PersistentVolume")
 
-				if err = v.createVolumeReplicationForPVC(pvc); err == nil {
-					continue
-				}
+			if err = v.uploadPV(pvc); err != nil {
+				log.Error(err, "failed to upload PV metadata")
+
+				requeue = true
+
+				// There's no point enabling PV data replication if
+				// replication of PV metadata to object store has failed.
+				break
 			}
 
-			log.Info("Requeuing due to failure in getting or creating VolumeReplication resource for PersistentVolumeClaim",
-				"errorValue", err)
+			log.Info("Creating VolumeReplication resource for PersistentVolumeClaim")
 
-			// requeue on failure to ensure any PVC not having a corresponding VR CR
-			requeue = true
+			if err = v.createVolumeReplicationForPVC(pvc); err == nil {
+				log.Error(err, "failed to upload PV metadata")
+
+				continue
+			}
 		}
+
+		log.Info("Requeuing due to failure in getting or creating VolumeReplication resource for PersistentVolumeClaim",
+			"errorValue", err)
+
+		// requeue on failure to ensure any PVC not having a corresponding VR CR
+		requeue = true
 	}
 
 	return requeue
 }
 
+// uploadPV checks if the VRG spec has been configured with an s3 endpoint,
+// validates the S3 endpoint, connects to it, gets the PV metadata of
+// the input PVC, creates a bucket in s3 store, upload's the PV metadata to
+// s3 store and downloads it for verification.  If an s3 endpoint is not
+// configured, then it assumes that VRG is running in a backup-less mode and
+// does not return an error, but logs a one-time warning.
+func (v *VRGInstance) uploadPV(pvc corev1.PersistentVolumeClaim) error {
+	// Create an s3 bucket name from the VRG's namespace and name, separating
+	// the two with a forward slash.  Note: VRG's namespace and name may also
+	// have dots or hyphens in their names.
+	vrgName := v.instance.Name
+	s3Bucket := v.instance.Namespace + "/" + vrgName
+
+	s3Endpoint := v.instance.Spec.S3Endpoint
+	if err := validateS3Endpoint(s3Endpoint, s3Bucket, v.log); err != nil {
+		return err
+	} else if s3Endpoint == "" {
```

Comment:
(future) Maybe at some point, we should treat an empty s3Endpoint as invalid.  What if the user forgot to set the s3Endpoint?  That would be an invalid configuration rather than backup-less mode. If, however, we will stick with this configuration requirement, then, maybe move the `else if s3Endpoint == ""` before the validation (before line 518) that way, you will not get rid of the `else` clause.

Reply:
Given that AVR is the user of VRG, if AVR forgot to set the s3Endpoint, it would be a bug in AVR, isn't it? :-) The need for an empty s3Endpoint is to support a potential use case of "backup-less" VRG operation mode.  

The reason it is coded in this admittedly peculiar way is to have to logic of one-time log.Warn message of running in backup-less mode inside the validateS3Endpoint method.  UploadPV is very close to 60 lines.